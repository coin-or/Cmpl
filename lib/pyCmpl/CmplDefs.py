#***********************************************************************
 #  This code is part of pyCMPL 
 #
 #  Copyright (C) 2013, 2014
 #  Mike Steglich - Technical University of Applied Sciences
 #  Wildau, Germany 
 #
 #  pyCMPL is a project of the Technical University of 
 #  Applied Sciences Wildau and the Institute for Operations Research 
 #  and Business Management at the Martin Luther University 
 #  Halle-Wittenberg.
 #  Please visit the project homepage <www.coliop.org>
 # 
 #  pyCMPL is free software; you can redistribute it and/or modify it 
 #  under the terms of the GNU Lesser General Public License as published by 
 #  the Free Software Foundation; either version 3 of the License, or 
 #  (at your option) any later version.
 # 
 #  pyCMPL is distributed in the hope that it will be useful, but WITHOUT
 #  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 #  or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public 
 #  License for more details.
 #  
 #  You should have received a copy of the GNU Lesser General Public License
 #  along with this program; if not, see <http://www.gnu.org/licenses/>.
 #
 #**********************************************************************

#!/usr/bin/python 
	
#*************** general definitions  ****************************
#don' change the order of the entries, because this would cause a lot of confusions

COMPATIBILITY = 2  # CMPL 1.10 

CMPL_UNKNOWN = 0
CMPL_OK = 1
CMPL_WARNINGS = 2
CMPL_FAILED = 3
SOLVER_OK = 4
SOLVER_FAILED = 5

CMPLSERVER_OK = 6
CMPLSERVER_ERROR = 7
CMPLSERVER_BUSY = 8
CMPLSERVER_CLEANED = 9 
CMPLSERVER_WARNING = 10

PROBLEM_RUNNING = 11
PROBLEM_FINISHED = 12
PROBLEM_CANCELED = 13 
PROBLEM_NOTRUNNING = 14 

CMPLGRID_SCHEDULER_UNKNOWN = 15
CMPLGRID_SCHEDULER_OK = 16
CMPLGRID_SCHEDULER_ERROR = 17
CMPLGRID_SCHEDULER_BUSY = 18
CMPLGRID_SCHEDULER_SOLVER_NOT_AVAILABLE = 19
CMPLGRID_SCHEDULER_WARNING = 20
CMPLGRID_SCHEDULER_PROBLEM_DELETED = 21

CMPL_GRID  = 22
STANDALONE_SERVER = 23
SERVER_UNKNOWN = 24

CMPL_STATUS_TXT = { CMPL_UNKNOWN    : "CMPL_UNKNOWN" , \
                    CMPL_OK         : "CMPL_OK", \
                    CMPL_WARNINGS   : "CMPL_WARNINGS ", \
                    CMPL_FAILED     : "CMPL_FAILED", \
                    SOLVER_OK       : "SOLVER_OK", \
                    SOLVER_FAILED   : "SOLVER_FAILED", \
                    CMPLSERVER_OK   : "CMPLSERVER_OK", \
                    CMPLSERVER_ERROR: "CMPLSERVER_ERROR", \
                    CMPLSERVER_BUSY : "CMPLSERVER_BUSY",  \
                    CMPLSERVER_CLEANED   : "CMPLSERVER_CLEANED", \
                    CMPLSERVER_WARNING   : "CMPLSERVER_WARNING ", \
                    PROBLEM_RUNNING : "PROBLEM_RUNNING", \
                    PROBLEM_FINISHED  : "PROBLEM_FINISHED", \
                    PROBLEM_CANCELED  : "PROBLEM_CANCELED", \
                    PROBLEM_NOTRUNNING  : "PROBLEM_NOTRUNNING", \
                    CMPLGRID_SCHEDULER_UNKNOWN  : "CMPLGRID_SCHEDULER_UNKNOWN", \
                    CMPLGRID_SCHEDULER_OK  : "CMPLGRID_SCHEDULER_OK", \
                    CMPLGRID_SCHEDULER_ERROR  : "CMPLGRID_SCHEDULER_ERROR", \
                    CMPLGRID_SCHEDULER_BUSY  : "CMPLGRID_SCHEDULER_BUSY", \
                    CMPLGRID_SCHEDULER_SOLVER_NOT_AVAILABLE  : "CMPLGRID_SCHEDULER_SOLVER_NOT_AVAILABLE", \
                    CMPLGRID_SCHEDULER_WARNING  : "CMPLGRID_SCHEDULER_WARNING", \
                    CMPLGRID_SCHEDULER_PROBLEM_DELETED  : "CMPLGRID_SCHEDULER_PROBLEM_DELETED", \
                    CMPL_GRID  : "CMPL_GRID", \
                    STANDALONE_SERVER  : "STANDALONE_SERVER", \
                    SERVER_UNKNOWN  : "SERVER_UNKNOWN"}


#*************** end general definitions  ************************

		

		