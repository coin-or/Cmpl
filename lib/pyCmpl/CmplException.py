#***********************************************************************
 #  This code is part of pyCMPL 
 #
 #  Copyright (C) 2013
 #  Mike Steglich - Technical University of Applied Sciences
 #  Wildau, Germany 
 #
 #  pyCMPL is a project of the Technical University of 
 #  Applied Sciences Wildau and the Institute for Operations Research 
 #  and Business Management at the Martin Luther University 
 #  Halle-Wittenberg.
 #  Please visit the project homepage <www.coliop.org>
 # 
 #  pyCMPL is free software; you can redistribute it and/or modify it 
 #  under the terms of the GNU Lesser General Public License as published by 
 #  the Free Software Foundation; either version 3 of the License, or 
 #  (at your option) any later version.
 # 
 #  pyCMPL is distributed in the hope that it will be useful, but WITHOUT
 #  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 #  or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public 
 #  License for more details.
 #  
 #  You should have received a copy of the GNU Lesser General Public License
 #  along with this program; if not, see <http://www.gnu.org/licenses/>.
 #
 #**********************************************************************

#!/usr/bin/python 

from CmplMsg import *
import HTMLParser
import cStringIO

#*************** CmplException ***********************************
class CmplException(Exception):
	
	def __init__(self, msg, msgList=None):
		self.__msgStr=cStringIO.StringIO()
		
		if not "pyCmpl error:" in msg:
			self.__msgStr.write("pyCmpl error: " +msg)
		else:	
			self.__msgStr.write(msg)
		
		if msgList!=None:
			for m in msgList:
				self.__msgStr.write("\n" + m.type + " in file " + m.file + " at line " + m.line + " : " + HTMLParser.HTMLParser().unescape(m.description))
	
	def __str__(self):
		return "pyCmpl error:"
	
	def __del__(self):
		self.__msgStr.close()
	
	@property
	def msg(self):
		return self.__msgStr.getvalue()
#*************** end CmplException *******************************


		
		
		
		

		