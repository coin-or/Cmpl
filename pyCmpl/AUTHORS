Mike Steglich - Technical University of Applied Sciences Wildau, Germany - mike.steglich@th-wildau.de

	