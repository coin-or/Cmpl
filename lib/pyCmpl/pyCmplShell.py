#***********************************************************************
 #  This code is part of pyCMPL 
 #
 #  Copyright (C) 2013
 #  Mike Steglich - Technical University of Applied Sciences
 #  Wildau, Germany 
 #
 #  pyCMPL is a project of the Technical University of 
 #  Applied Sciences Wildau and the Institute for Operations Research 
 #  and Business Management at the Martin Luther University 
 #  Halle-Wittenberg.
 #  Please visit the project homepage <www.coliop.org>
 # 
 #  pyCMPL is free software; you can redistribute it and/or modify it 
 #  under the terms of the GNU Lesser General Public License as published by 
 #  the Free Software Foundation; either version 3 of the License, or 
 #  (at your option) any later version.
 # 
 #  pyCMPL is distributed in the hope that it will be useful, but WITHOUT
 #  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 #  or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public 
 #  License for more details.
 #  
 #  You should have received a copy of the GNU Lesser General Public License
 #  along with this program; if not, see <http://www.gnu.org/licenses/>.
 #
 #**********************************************************************

#!/usr/bin/python -i

import os
import sys
from pyCmpl import *

def pyCmplShell():
	if sys.version_info[:2] == (2,7):

		
		print "CMPL\'s Python API"
	
		sys.ps1 = "pyCmpl> "
		sys.ps2 = "....... "
	
		pyCmplAnswer=""
	
		while pyCmplAnswer!="quit()" and pyCmplAnswer!="exit()" :
			try:
				pyCmplAnswer = raw_input("pyCmpl> ")
				if pyCmplAnswer!="quit()" and pyCmplAnswer!="exit()":
					exec(pyCmplAnswer) 
			except CmplException, e:
				print e.msg
			except:
				print "Python error : " +str(sys.exc_info()[1])
	else:
		print "ERROR - pyCmpl only supports Python 2.7"
		
	exit(0)

if __name__ == "__main__":
	pyCmplShell()		
		
		
		

		