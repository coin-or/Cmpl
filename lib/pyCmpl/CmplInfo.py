#***********************************************************************
 #  This code is part of pyCMPL 
 #
 #  Copyright (C) 2013
 #  Mike Steglich - Technical University of Applied Sciences
 #  Wildau, Germany 
 #
 #  pyCMPL is a project of the Technical University of 
 #  Applied Sciences Wildau and the Institute for Operations Research 
 #  and Business Management at the Martin Luther University 
 #  Halle-Wittenberg.
 #  Please visit the project homepage <www.coliop.org>
 # 
 #  pyCMPL is free software; you can redistribute it and/or modify it 
 #  under the terms of the GNU Lesser General Public License as published by 
 #  the Free Software Foundation; either version 3 of the License, or 
 #  (at your option) any later version.
 # 
 #  pyCMPL is distributed in the hope that it will be useful, but WITHOUT
 #  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 #  or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public 
 #  License for more details.
 #  
 #  You should have received a copy of the GNU Lesser General Public License
 #  along with this program; if not, see <http://www.gnu.org/licenses/>.
 #
 #**********************************************************************

#!/usr/bin/python 


#import os
import re
import cStringIO 

#from CmplDefs import *
from CmplException import *


#*************** CmplInfo *****************************************
class CmplInfo(object):
		
	#*********** constructor ******
	def __init__(self, cmplInfo=None):
		self.__cmplInfoFile=cmplInfo
		self.__instanceName=""
		self.__varProdFileName = ""
		self.__varProdText = ""
		self.__statisticsFileName = ""
		self.__statisticsText = ""
		self.__matrixFileName = ""
		self.__matrixText = ""
	#*********** end constructor ******
	
	# getter and setter 	
	@property
	def varProdFileName(self):
		return self.__varProdFileName
	
	@property
	def varProdText(self):
		return self.__varProdText
		
	@property
	def statisticsFileName(self):
		return self.__statisticsFileName
		
	@property
	def statisticsText(self):
		return self.__statisticsText
		
	@property
	def matrixFileName(self):
		return self.__matrixFileName
		
	@property
	def matrixText(self):
		return self.__matrixText
	#end getter and setter
	
		
	#*********** readCmplInfo ********	
	def readCmplInfo(self, cmplInfoStr=None):
	
		if cmplInfoStr == None and self.__cmplInfoFile == None:
			raise CmplException("Neither cmplInfoFile nor cmplInfoString defined")
		
		if 	self.__cmplInfoFile != None:
			if not os.path.isfile(self.__cmplInfoFile):
				raise CmplException("Can't read the CmplInfo file: " + self.__cmplInfoFile  )
	
			try:	
				f = open(self.__cmplInfoFile, "r")
				cmplInfoStr = f.read()
				f.close()
			except IOError, e:
				raise CmplServerException("IO error for CmplInfo file: "+self.__cmplInfoFile)
				
				
		lines = cStringIO.StringIO(cmplInfoStr) 
		
		lineNr=1
		generalSection=False
		varProdSection=False
		statSection=False
		matrixSection=False
			
		for line in lines:
			
			if lineNr == 1:
				if not "<?xml version" in line: 
					raise CmplException("File - is not a XML file!")
				else:
					lineNr += 1
					continue
			if lineNr == 2:
				if not "<CmplInfo" in line: 
					raise CmplException("Cant't read CmplInfo file - wrong file type!")
				else:
					lineNr += 1
					continue
	
			if "<general>" in line:
				generalSection=True
				varProdSection=False
				statSection=False
				matrixSection=False
				continue
			if "</general>" in line:
				generalSection=False
				continue
			if "<statistics" in line:
				self.__statisticsFileName=re.findall("\"([^\"]*)\"", line)[0]
				generalSection=False
				varProdSection=False
				statSection=True
				matrixSection=False
				continue
			if "</statistics" in line:
				statSection=False
				continue
			if "<variableProducts" in line:
				self.__varProdFileName=re.findall("\"([^\"]*)\"", line)[0]
				generalSection=False
				varProdSection=True
				statSection=False
				matrixSection=False
				continue
			if "</variableProducts" in line:
				varProdSection=False
				continue
			if "<matrix" in line:
				self.__matrixFileName=re.findall("\"([^\"]*)\"", line)[0]
				generalSection=False
				varProdSection=False
				statSection=False
				matrixSection=True
				continue
			if "</matrix" in line:
				matrixSection=False
				continue
				
			if generalSection:
				if "<generalStatus" in line:
					self.__instanceName = re.findall("<instanceName>([^\"]*)</instanceName>", line)[0]
					continue
		
			if statSection:
				self.__statisticsText=self.__statisticsText+line
		
			if varProdSection:
				self.__varProdText=self.__varProdText+line
				
			if matrixSection:
				self.__matrixText=self.__matrixText+line
							
	#*********** end readCmplInfo ****	
		
	

#*************** CmplInfo *****************************************		
		
		
		

		