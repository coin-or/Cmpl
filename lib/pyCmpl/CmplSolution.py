#***********************************************************************
 #  This code is part of pyCMPL 
 #
 #  Copyright (C) 2013
 #  Mike Steglich - Technical University of Applied Sciences
 #  Wildau, Germany 
 #
 #  pyCMPL is a project of the Technical University of 
 #  Applied Sciences Wildau and the Institute for Operations Research 
 #  and Business Management at the Martin Luther University 
 #  Halle-Wittenberg.
 #  Please visit the project homepage <www.coliop.org>
 # 
 #  pyCMPL is free software; you can redistribute it and/or modify it 
 #  under the terms of the GNU Lesser General Public License as published by 
 #  the Free Software Foundation; either version 3 of the License, or 
 #  (at your option) any later version.
 # 
 #  pyCMPL is distributed in the hope that it will be useful, but WITHOUT
 #  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 #  or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public 
 #  License for more details.
 #  
 #  You should have received a copy of the GNU Lesser General Public License
 #  along with this program; if not, see <http://www.gnu.org/licenses/>.
 #
 #**********************************************************************

#!/usr/bin/python 

import os
import re
import cStringIO

from CmplException import *

#*************** CmplSolElement *************************************
class CmplSolElement(object):
	#*********** constructor **********
	def __init__(self):
		self.__idx = 0
		self.__name = ""
		self.__type = ""
		self.__activity = 0
		self.__lowerBound = 0
		self.__upperBound = 0
		self.__hasMarginal = False
		self.__marginal = 0
	#*********** end constructor ******
	
	# getter and setter 
	@property
	def idx(self):
		return self.__idx 
		
	def setIdx(self, idx):
		self.__idx = idx
	
	@property
	def name(self):
		return self.__name 
	
	def setName(self, name):
		self.__name = name
	
	@property
	def type(self):	
		return self.__type 
		
	def setType(self, type):
		self.__type=type
		
	@property
	def activity(self):
		return self.__activity 
		
	def setActivity(self, activity):
		self.__activity=activity
		
	@property
	def lowerBound(self):		
		return self.__lowerBound
		
	def setLowerBound(self, lB):
		self.__lowerBound = lB
		
	@property
	def upperBound(self):
		return self.__upperBound 
		
	def setUpperBound(self, uB):
		self.__upperBound = uB
		
	@property
	def marginal(self):
		return self.__marginal 
		
	def setMarginal(self, marginal):
		self.__marginal=marginal
	# end getter and setter 
	
		
#*************** end CmplSolElement ******************************

#*************** CmplSolution  ***********************************
class CmplSolution(object):

	#*********** constructor **********
	def __init__(self):
		self.__idx = 0
		self.__status = ""
		self.__value = 0
		self.__variables =[]
		self.__constraints = []
	#*********** end constructor ******
	
	# getter and setter 
	@property
	def idx(self):
		return self.__idx 
		
	def setIdx(self, idx):
		self.__idx = idx
	
	@property
	def status(self):
		return self.__status
		
	def setStatus(self, status):
		self.__status=status
		
	@property
	def value(self):
		return self.__value
		
	def setValue(self, val):
		self.__value=val
	
	@property
	def variables(self):
		return self.__variables
		
	def setVar(self, var):
		self.__variables.append(var)
		
	@property
	def constraints(self):
		return self.__constraints
		
	def setCon(self, con):
		self.__constraints.append(con)
		
	# end getter and setter 	
		
#*************** end CmplSolution ********************************


#*************** CmplSolutions  **********************************
class CmplSolutions(object):

	#*********** constructor **********
	def __init__(self, solFile=None):
		self.__nrOfVariables = 0
		self.__nrOfConstraints = 0
		self.__objectiveName = ""
		self.__objectiveSense = ""
		self.__nrOfSolutions = 0
		self.__solverName = ""
		self.__solverMsg = ""
		self.__varDisplayOptions = ""
		self.__conDisplayOptions = ""
		self.__asciiSolFile = ""
		self.__csvSolFile = "" 
		self.__cmplSolFile = ""
		self.__solList = []
		self.__solutionFile = solFile
		self.__solFileContent = ""
		self.__isIntegerProgram = False
	#*********** end constructor ******
	
	# getter
	@property
	def nrOfVariables(self):
		return self.__nrOfVariables
		
	@property
	def nrOfConstraints(self):
		return self.__nrOfConstraints 
		
	@property
	def nrOfConstraints(self):
		return self.__nrOfConstraints 
		
	@property
	def objectiveName(self):
		return self.__objectiveName 
		
	@property
	def objectiveSense(self):
		return self.__objectiveSense 
		
	@property
	def nrOfSolutions(self):
		return self.__nrOfSolutions 
		
	@property
	def solver(self):
		return self.__solverName 
		
	@property
	def solverMessage(self):
		return self.__solverMsg 
		
	@property
	def varDisplayOptions(self):
		return self.__varDisplayOptions 
		
	@property
	def conDisplayOptions(self):
		return self.__conDisplayOptions
		
	@property
	def asciiSolFile(self):
		return self.__asciiSolFile
		
	@property
	def csvSolFile(self):
		return self.__csvSolFile
			
	@property
	def cmplSolFile(self):
		return self.__cmplSolFile
		
	@property
	def solutions(self):
		return self.__solList
		
	@property
	def solution(self):
		return self.__solList[0]
		
	@property
	def solFileContent(self):
		return self.__solFileContent
	
	def delSolFileContent(self):
		self.__solFileContent=""
	
	@property
	def isIntegerProgram(self):
		return self.__isIntegerProgram
	
	#*********** readSolLine *************	
	def _readSolLine(self, line):
	
		tmpList=re.findall("\"([^\"]*)\"", line)

		x = CmplSolElement()
		x.setIdx(int(tmpList[0]))
		x.setName(tmpList[1])
		x.setType(tmpList[2])
		
		isSolverStudio=False
		try:		
			isSolverStudio = os.environ['SOLVERSTUDIO']=='1'
		except:
			pass
		
		if "<constraint" in line:
			x.setActivity(float(tmpList[3]))
		else:
			if x.type=='I' or x.type=='B':
				self.__isIntegerProgram=True
				
				if not isSolverStudio:	
					x.setActivity(long(tmpList[3]))
				else:
					x.setActivity(int(tmpList[3]))
			else:
				x.setActivity(float(tmpList[3]))
		
		if not isSolverStudio:	
			x.setLowerBound(float(tmpList[4]))
			x.setUpperBound(float(tmpList[5]))
		else:
			if tmpList[4].upper()=='-INF':
				x.setLowerBound(-1e+100)
			else:
				x.setLowerBound(float(tmpList[4]))
			if tmpList[5].upper()=='INF':	
				x.setUpperBound(1e+100)
			else:
				x.setUpperBound(float(tmpList[5]))

		#if tmpList[6]=='-':
		if tmpList[6]=='NaN':
			x.setMarginal(float('NaN'))
		else: 
			x.setMarginal(float(tmpList[6]))
		return x
	
	#*********** end readSolLine *********	
	
	
	#*********** readSolution ************	
	def readSolution(self, solStr=None):
		
		if (solStr == None or len(solStr)==0) and self.__solutionFile == None:
			raise CmplException("No solution found")
		
		if self.__solutionFile != None:
			if not os.path.isfile(self.__solutionFile):
				raise CmplException("No solution found" )
				
			try:	
				f = open(self.__solutionFile, "r")
				solStr = f.read()
				f.close()
			except IOError, e:
				raise CmplServerException("IO error for solution file ")
		
		lines = cStringIO.StringIO(solStr) 
		
		lineNr=1
		generalSection=False
		solutionSection=False
		varSection=False
		conSection=False
		s = CmplSolution()
		self.__solFileContent=solStr
		
		for line in lines:
			
			if lineNr == 1:
				if not "<?xml version" in line: 
					raise CmplException("File " + self.__solutionFile + " - is not a XML file!")
				else:
					lineNr += 1
					continue
			if lineNr == 2:
				if not "<CmplSolutions" in line: 
					raise CmplException("Cant't read cmplSolution file " + self.__solutionFile + " - wrong file type!")
				else:
					lineNr += 1
					continue
			
			if "<general>" in line:
				generalSection=True
				solutionSection=False
				continue
			if "</general>" in line:
				generalSection=False
				continue
				
			if "<solution" in line:
				generalSection=False
				solutionSection=True
				s = CmplSolution()
				tmpList = re.findall("\"([^\"]*)\"", line)
				s.setIdx(int(tmpList[0]))
				s.setStatus(tmpList[1])
				s.setValue(float(tmpList[2]))
				continue
			if "</solution" in line:
				solutionSection=False
				self.__solList.append(s)
				continue
				
			if generalSection:
				if "<nrOfVariables" in line:
					self.__nrOfVariables = int(re.findall("<nrOfVariables>([^\"]*)</nrOfVariables>", line)[0])
					continue
				if "<nrOfConstraints" in line:
					self.__nrOfConstraints = int(re.findall("<nrOfConstraints>([^\"]*)</nrOfConstraints>", line)[0])
					continue
				if "<objectiveName" in line:
					self.__objectiveName = re.findall("<objectiveName>([^\"]*)</objectiveName>", line)[0]
					continue
				if "<objectiveSense" in line:
					self.__objectiveSense = re.findall("<objectiveSense>([^\"]*)</objectiveSense>", line)[0]
					continue
				if "<nrOfSolutions" in line:
					self.__nrOfSolutions = int(re.findall("<nrOfSolutions>([^\"]*)</nrOfSolutions>", line)[0])
					continue
				if "<solverName" in line:
					self.__solverName = re.findall("<solverName>([^\"]*)</solverName>", line)[0]
					continue
				if "<solverMsg" in line:
					self.__solverMsg = re.findall("<solverMsg>([^\"]*)</solverMsg>", line)[0]
					continue
				if "<variablesDisplayOptions" in line:
					self.__varDisplayOptions = re.findall("<variablesDisplayOptions>([^\"]*)</variablesDisplayOptions>", line)[0]
					continue
				if "<constraintsDisplayOptions" in line:
					self.__conDisplayOptions = re.findall("<constraintsDisplayOptions>([^\"]*)</constraintsDisplayOptions>", line)[0]
					continue
				if "<asciiSolFile" in line:
					self.__asciiSolFile = re.findall("<asciiSolFile>([^\"]*)</asciiSolFile>", line)[0]
					continue
				if "<csvSolFile" in line:
					self.__csvSolFile = re.findall("<csvSolFile>([^\"]*)</csvSolFile>", line)[0]
					continue
				if "<cmplSolFile" in line:
					self.__cmplSolFile = re.findall("<cmplSolFile>([^\"]*)</cmplSolFile>", line)[0]
					continue	
				
			
			if solutionSection:
				if "<variables>" in line:
					varSection=True
					conSection=False
					continue
				if "</variables>" in line:
					varSection=False
					continue
				if "<linearConstraints>" in line:
					varSection=False
					conSection=True
					continue
				if "</linearConstraints>" in line:
					conSection=False
					continue
				if varSection:
					if "<variable" in line:
						s.setVar(self._readSolLine(line))
						continue
				if conSection:
					if "<constraint" in line:
						s.setCon(self._readSolLine(line))
						continue
		
	#*********** send readSolution *******




#*************** end CmplSolutions  ******************************

		