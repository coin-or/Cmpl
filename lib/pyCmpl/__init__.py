from CmplDefs import *
from Cmpl import *
from CmplException import *
from CmplMsg import *
from CmplTools import *
from CmplSolution import *
from CmplSet import *
from CmplParameter import *
from CmplInstance import *
from CmplInfo import *

if not sys.version_info[:2] == (2,7):
	print "ERROR - pyCmpl only supports Python 2.7"
	exit(-1)
	