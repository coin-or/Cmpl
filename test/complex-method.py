#***********************************************************************
 #  This code is part of CMPL
 #
 #  Copyright (C) 2007, 2008, 2009, 2010, 2011, 2012, 2013
 #  Thomas Schleiff - Halle(Saale), Germany and
 #  Mike Steglich - Technical University of Applied Sciences
 #  Wildau, Germany 
 #
 #  CMPL is a project of the Technical University of 
 #  Applied Sciences Wildau and the Institute for Operations Research 
 #  and Business Management at the Martin Luther University 
 #  Halle-Wittenberg.
 #  Please visit the project homepage <www.coliop.org>
 # 
 #  CMPL is free software; you can redistribute it and/or modify it 
 #  under the terms of the GNU General Public License as published by 
 #  the Free Software Foundation; either version 3 of the License, or 
 #  (at your option) any later version.
 # 
 #  CMPL is distributed in the hope that it will be useful, but WITHOUT
 #  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 #  or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public 
 #  License for more details.
 #  
 #  You should have received a copy of the GNU General Public License
 #  along with this program; if not, see <http://www.gnu.org/licenses/>.
 #
 #**********************************************************************

#!/usr/bin/python 

from pyCmpl import *

try: 
	prodmix = Cmpl("production-mix.cmpl")
	prodmixSub = Cmpl("production-mix-substitute.cmpl")
	
	products = CmplSet("products")
	products.setValues(1,5)
	
	machines = CmplSet("machines")
	machines.setValues(1,4)
	
	c = CmplParameter("c",products)
	c.setValues([180,540,300,410,620])
	
	a = CmplParameter("a",machines, products)
	a.setValues( [[8,64,40,46,63],[20,40,35,50,70],[30,20,56,12,112],[25,32,28,4,48]])
	
	b = CmplParameter("b", machines)
	b.setValues([20000,25000,10000,15000])
	
	xMax = CmplParameter("xMax", products)
	xMax.setValues([300,160, 1e10, 200, 1e10])
	
	maxDeltaA = CmplParameter("maxDeltaA", machines, products)
	
	prodmix.setSets(machines,products)
	prodmix.setParameters(c,a,b,xMax)
	
	prodmix.solve()
	
	deltaAUpperBounds = [[2,15,0,0,0], [10,10,10,10,10], [10,8,0,0,0],[10,10,10,10,10]]

	maxDeltaAMatrix =[]
	restrictedMachines=[]
	i = 0
	for r in prodmix.solution.constraints:
		if r.marginal>0:
			maxDeltaAMatrix.append(deltaAUpperBounds[i])
		else:
			maxDeltaAMatrix.append([ 0 for j in range(products.len) ] )
		i = i+1
	maxDeltaA.setValues(maxDeltaAMatrix)
			
	prodmixSub.setSets(machines,products)
	prodmixSub.setParameters(c,a,b,xMax,maxDeltaA )
	
	prodmixSub.solve()
	prodmixSub.varByName()
	
	newA = [[0,0,0,0,0], [0,0,0,0,0], [0,0,0,0,0],[0,0,0,0,0]]
	for i in machines.values:
		for j in products.values:
			if prodmixSub.w[i].activity != 0:
				newA[i-1][j-1] = a.values[i-1][j-1]-deltaAUpperBounds[i-1][j-1]* \
								(prodmixSub.w[i].activity/(prodmixSub.s[i].activity+prodmixSub.w[i].activity))
			else:
				newA[i-1][j-1] = a.values[i-1][j-1]
		
	print "%-10s%20s%20s" % ("","original model", "modified model")
	print "%-10s" % ("Variables")
	for v in prodmix.solution.variables:
		i = prodmix.solution.variables.index(v)
		print "%10s%18.2f%18.2f" % (v.name, v.activity, prodmixSub.solution.variables[i].activity )
	
	print "%-10s" % ("Constraints")
	for r in prodmix.solution.constraints:
		i = prodmix.solution.constraints.index(r)
		print "%10s%18.2f%18.2f" % (r.name, r.activity, prodmixSub.solution.constraints[i].activity )
	
	print "\n\n%20s" % ("Original coefficient matrix")
	for i in machines.values:
		row = "%10s" % prodmix.solution.constraints[i-1].name
		for j in products.values:
			row = row + "%8.2f" % a.values[i-1][j-1]
		print row 
	
	print "\n%20s" % ("Modified coefficient matrix")	
	for i in machines.values:
		row = "%10s" % prodmix.solution.constraints[i-1].name
		for j in products.values:
			row = row + "%8.2f" % newA[i-1][j-1]
		print row 
		
	
except CmplException, e:
	print e.msg
	
	