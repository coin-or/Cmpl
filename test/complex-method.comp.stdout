                original model      modified model
Variables 
      x[1]            153.64            284.00
      x[2]            149.54            160.00
      x[3]              0.00              0.00
      x[4]            200.00            200.00
      x[5]              0.00              0.00
Constraints
    res[1]          20000.00          18744.00
    res[2]          19054.50          22080.00
    res[3]          10000.00          10000.00
    res[4]           9426.36          13020.00


Original coefficient matrix
    res[1]    8.00   64.00   40.00   46.00   63.00
    res[2]   20.00   40.00   35.00   50.00   70.00
    res[3]   30.00   20.00   56.00   12.00  112.00
    res[4]   25.00   32.00   28.00    4.00   48.00

Modified coefficient matrix
    res[1]    6.00   49.00   40.00   46.00   63.00
    res[2]   20.00   40.00   35.00   50.00   70.00
    res[3]   20.00   12.00   56.00   12.00  112.00
    res[4]   25.00   32.00   28.00    4.00   48.00
